### PR TITLE
Option to use JsonWebTokenHandler in oidc handler

### PR DESCRIPTION
**Pull request by keegan-caruso**
# Option to use JsonWebTokenHandler in oidc handler

## Description

Gives the option to use JsonWebTokenHandler instead of JwtSecurityTokenHandler in OIDC Handler. This gives an async model, last known good support for metadata, and better performance.

### CODE REVIEW

File: eng/Versions.props [hunk 1]
```diff
@@ -248,15 +248,15 @@
     <MicrosoftCodeAnalysisCSharpAnalyzerTestingXUnitVersion>1.1.2-beta1.22531.1</MicrosoftCodeAnalysisCSharpAnalyzerTestingXUnitVersion>
     <MicrosoftCodeAnalysisCSharpCodeFixTestingXUnitVersion>1.1.2-beta1.22531.1</MicrosoftCodeAnalysisCSharpCodeFixTestingXUnitVersion>
     <MicrosoftCssParserVersion>1.0.0-20230414.1</MicrosoftCssParserVersion>
-    <MicrosoftIdentityModelLoggingVersion>6.15.1</MicrosoftIdentityModelLoggingVersion>
-    <MicrosoftIdentityModelProtocolsOpenIdConnectVersion>6.15.1</MicrosoftIdentityModelProtocolsOpenIdConnectVersion>
-    <MicrosoftIdentityModelProtocolsWsFederationVersion>6.15.1</MicrosoftIdentityModelProtocolsWsFederationVersion>
+    <MicrosoftIdentityModelLoggingVersion>7.0.0-preview</MicrosoftIdentityModelLoggingVersion>
```
Comment (eerhardt) [inline eng/Versions.props:251]: Will `7.0.0` stable be released before .NET 8 ships?

File: src/Security/Authentication/OpenIdConnect/src/LoggingExtensions.cs [hunk 2]
```diff
@@ -164,4 +164,10 @@ internal static partial class LoggingExtensions
     [LoggerMessage(55, LogLevel.Error, "The remote signout request was ignored because the 'iss' parameter didn't match " +
                          "the expected value, which may indicate an unsolicited logout.", EventName = "RemoteSignOutIssuerInvalid")]
     public static partial void RemoteSignOutIssuerInvalid(this ILogger logger);
+
+    [LoggerMessage(56, LogLevel.Error, "Unable to validate the 'id_token', no suitable TokenHandler was found for: '{IdToken}'.", EventName = "UnableToValidateIdTokenFromHandler")]
+    public static partial void UnableToValidateIdTokenFromHandler(this ILogger logger, string idToken);
+
+    [LoggerMessage(57, LogLevel.Error, "The Validated Security Token must be of type JsonWebToken, but instead its type is: '{SecurityTokenType}'", EventName = "InvalidSecurityTokenTypeFromHandler")]
+    public static partial void InvalidSecurityTokenTypeFromHandler(this ILogger logger, string? securityTokenType);
```
Comment (eerhardt) [inline src/Security/Authentication/OpenIdConnect/src/LoggingExtensions.cs:172]:
```suggestion
    public static partial void InvalidSecurityTokenTypeFromHandler(this ILogger logger, string securityTokenType);
```

I don't think we'd want `null` to be passed here. If the object is `null`, we should probably say it is `null` somehow rather than have a message that says `"but instead its type is: "`.

File: eng/Versions.props (hunk 1 above)
Comment (jennyf19) [reply eng/Versions.props:251]: Yes, definitely. We can ship out of preview by end of August, at the latest. Depends on perf/stress testing and customer feedback.

File: src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs [hunk 3]
```diff
@@ -717,7 +729,19 @@ protected override async Task<HandleRequestResult> HandleRemoteAuthenticateAsync
 
                 // At least a cursory validation is required on the new IdToken, even if we've already validated the one from the authorization response.
                 // And we'll want to validate the new JWT in ValidateTokenResponse.
-                var tokenEndpointUser = ValidateToken(tokenEndpointResponse.IdToken, properties, validationParameters, out var tokenEndpointJwt);
+                ClaimsPrincipal tokenEndpointUser;
+                JwtSecurityToken tokenEndpointJwt;
+
+                if (Options.UseTokenHandler)
+                {
+                    var tokenValidationResult = await ValidateTokenUsingHandlerAsync(tokenEndpointResponse.IdToken, properties, validationParameters);
```
Comment (brentschmaltz) [inline src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs:737]: We need to check if the tokenValidationResult is valid as this does not throw.

File: src/Security/Authentication/OpenIdConnect/src/OpenIdConnectOptions.cs [hunk 4]
```diff
@@ -38,6 +40,10 @@ public OpenIdConnectOptions()
         SignedOutCallbackPath = new PathString("/signout-callback-oidc");
         RemoteSignOutPath = new PathString("/signout-oidc");
         SecurityTokenValidator = _defaultHandler;
+        TokenHandler = _defaultTokenHandler;
```
Comment (jennyf19) [inline src/Security/Authentication/OpenIdConnect/src/OpenIdConnectOptions.cs:43]: Should `TokenHandler` be plural, as in the WsFed and JwtBearer PRs? Or all singular? @keegan-caruso

File: src/Security/Authentication/OpenIdConnect/src/OpenIdConnectOptions.cs [hunk 5]
```diff
@@ -255,6 +261,14 @@ public new OpenIdConnectEvents Events
     /// </summary>
     public ISecurityTokenValidator SecurityTokenValidator { get; set; }
 
+    /// <summary>
+    /// Gets or sets the <see cref="TokenHandler"/> used to validate identity tokens.
+    /// <para>
+    /// This will be used instead of <see cref="SecurityTokenValidator"/> if <see cref="UseTokenHandler"/> is <see langword="true"/>
+    /// </para>
+    /// </summary>
+    public TokenHandler TokenHandler { get; set; }
```
Comment (jennyf19) [inline src/Security/Authentication/OpenIdConnect/src/OpenIdConnectOptions.cs:270]: The other PR returns `public IList<TokenHandler> TokenHandlers { get; private set; }` should everything be aligned or does that matter?

File: src/Security/Authentication/OpenIdConnect/samples/OpenIdConnectSample/Startup.cs [hunk 6]
```diff
@@ -266,7 +266,7 @@ await context.SignOutAsync(OpenIdConnectDefaults.AuthenticationScheme, new Authe
                     // Persist the new acess token
                     props.UpdateTokenValue("access_token", payload.RootElement.GetString("access_token"));
                     props.UpdateTokenValue("refresh_token", payload.RootElement.GetString("refresh_token"));
-                    if (payload.RootElement.TryGetProperty("expires_in", out var property) && property.TryGetInt32(out var seconds))
+                    if (payload.RootElement.TryGetProperty("expires_in", out var property) && int.TryParse(property.GetString(), out var seconds))
```
Comment (davidfowl) [inline src/Security/Authentication/OpenIdConnect/samples/OpenIdConnectSample/Startup.cs:269]: Why this change?

File: src/Security/Authentication/OpenIdConnect/src/OpenIdConnectOptions.cs (hunk 4 above)
Comment (jennyf19) [reply src/Security/Authentication/OpenIdConnect/src/OpenIdConnectOptions.cs:43]: we discussed off-line. resolved.

File: src/Security/Authentication/OpenIdConnect/src/OpenIdConnectOptions.cs (hunk 5 above)
Comment (jennyf19) [reply src/Security/Authentication/OpenIdConnect/src/OpenIdConnectOptions.cs:270]: resolved

File: src/Security/Authentication/OpenIdConnect/samples/OpenIdConnectSample/Startup.cs (hunk 6 above)
Comment (keegan-caruso) [reply src/Security/Authentication/OpenIdConnect/samples/OpenIdConnectSample/Startup.cs:269]: property.TryGetInt32 was throwing a InvalidOperationException on ValueKind not being a number

Comment (eerhardt) [reply src/Security/Authentication/OpenIdConnect/samples/OpenIdConnectSample/Startup.cs:269]: https://github.com/dotnet/runtime/blob/0f56e166b16100c23dc81ae082f6155362b7c596/src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs#L481

is the check that ensures it is a Number.

This seems like a bug. There should be no reason for a `TryGet` method to throw an exception.

File: src/Security/Authentication/OpenIdConnect/src/LoggingExtensions.cs (hunk 2 above)
Comment (keegan-caruso) [reply src/Security/Authentication/OpenIdConnect/src/LoggingExtensions.cs:172]: This is the same approach as taken in InvalidSecurityTokenType earlier in the file
